File: java-maven/src/main/java/javamaven/javamaven/service/UserService.java
package javamaven.javamaven.service;

import javamaven.javamaven.entity.User;
import javamaven.javamaven.exception.BadRequestException;
import javamaven.javamaven.exception.NotFoundException;
import javamaven.javamaven.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    public UserService(UserRepository userRepository) { this.userRepository = userRepository; }

    public List<User> getAllUsers() { return (List<User>) userRepository.findAll(); }

    public User getUser(Long id) {
        Optional<User> userById = userRepository.findById(id);
        if (userById.isEmpty()) throw new NotFoundException("User not found");
        return userById.get();
    }

    public User addNewUser(User user) {
        Optional<User> userByEmail = userRepository.findUserByEmail(user.getEmail());
        if (userByEmail.isPresent()) {
            throw new BadRequestException("User already exist");
        }
        return userRepository.save(user);
    }

    public void userSignIn(String email, String password) {
        Optional<User> userByEmail = userRepository.findUserByEmail(email);
        if (userByEmail.isEmpty()){
            throw new BadRequestException("Please check your email and password");
        }
        if (!userByEmail.get().getPassword().equals(password)) {
            throw new BadRequestException("Please check your email and password");
        }
    }

    public User userChangePassword(Long id, String oldPassword, String newPassword) {
        Optional<User> userById = userRepository.findById(id);
        if (userById.isEmpty()) throw new BadRequestException("User does not exist");
        if (userById.get().getPassword() != oldPassword) throw new BadRequestException("Old password does not match");
        userById.get().setPassword(newPassword);
        userRepository.save(userById.get());
        return userById.get();
    }
}

File: kotlin-maven/src/main/kotlin/kotlinholidaytracker/kotlinholidaytracker/service/UserService.kt
package kotlinholidaytracker.kotlinholidaytracker.service

import kotlinholidaytracker.kotlinholidaytracker.entity.User
import kotlinholidaytracker.kotlinholidaytracker.exception.BadRequestException
import kotlinholidaytracker.kotlinholidaytracker.exception.NotFoundException
import kotlinholidaytracker.kotlinholidaytracker.repository.UserRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserService (private val userRepository: UserRepository) {
  fun getAllUsers(): List<User> {
    return userRepository.findAll()
  }

  fun getUser(id: Long): User {
    val userEntry: Optional<User> = userRepository.findById(id)
    if (userEntry.isEmpty) throw NotFoundException("User not found")
    return userEntry.get()
  }

  fun addNewUser(user: User): User {
    val userByEmail: Optional<User> = userRepository.findUserByEmail(user.email)
    if (userByEmail.isPresent) throw BadRequestException("User already exist")
    return userRepository.save(user)
  }

  fun userSignIn(email: String, password: String) {
    val userEntry: Optional<User> = userRepository.findUserByEmail(email)
    if (userEntry.isEmpty) throw BadRequestException("Please check your email and password")
    if (userEntry.get().password != password) throw BadRequestException("Please check your email and password")
  }

  fun userChangePassword(id: Long, oldPassword: String, newPassword: String): User {
    val userById: Optional<User> = userRepository.findById(id)
    if (userById.isEmpty) throw BadRequestException("User does not exist")
    if (userById.get().password != oldPassword) throw BadRequestException("Old password does not match")
    userById.get().password = newPassword
    userRepository.save(userById.get())
    return userById.get()
  }
}